### DISCUSSION TITLE
Connector Timeout Settings: Does Zero = No Timeout/Indefinite Wait?

### DISCUSSION THREAD

**Discussion by flugruger** [nextgenhealthcare/connect#5066] [Q&A]
Could you help with a question about Mirth connector timeout settings? I've been told some of our connectors have zero entered in the timeout settings because this disables the timeout, allowing the connector to wait indefinitely. 

_Regardless of why someone may have needed that or if that is a good idea,_ 
**Can you tell me if a zero ms timeout setting actually does disable the timeout?**

Thanks!

**Comment by jonbartels** [comment]
1. What does the manual say?
2. What do the tooltips on the fields in the UI say?
3. What connector types are you interested in?

This may require scrounging through code to validate the behavior for "0". A good search to start with is https://github.com/nextgenhealthcare/connect/search?l=Java&p=2&q=timeout then look for the actual connector implementations in the server or connector code. You can safely ignore the client code, that's just the UI. You're looking for accessors for the timeouts that check them for zero. You are also looking for where the timeout is passed to another library then you have to interpret that libraries usage of a `0` timeout.

Some examples of why this is hard to answer as a universal behavior:

HTTP Sender says:
> Sets the socket timeout (SO_TIMEOUT) in milliseconds to be used executing the method. A timeout value of zero is interpreted as an infinite timeout.

TCP Listener says this, but it defaults to zero:
> The amount of time, in milliseconds, to wait without receiving a message before closing a connection.

This text DOES NOT specify if that is an infinite timeout or if it means the connection will wait for the end of message sequence. The TCP listener also has remarks about timeouts on "Keep Connection Open". I think this highlights that MC does not have a universal behavior for all timeout fields. 

Finally - In my opinion an infinite timeout is not a good practice. At least not a good practice without a genuine connection pool manager. Large timeouts are fine, but you eventually want to terminate a connection that is not being used.

I hope I am helping with this kind of answer. If you want a more definite answer, narrow your focus to a few connectors that you're actually using. If you are setting infinite timeouts to solve a specific problem, describe that problem.

**Reply by flugruger** [reply]
@jonbartels ... thanks! 
> I hope I am helping with this kind of answer. If you want a more definite answer, narrow your focus to a few connectors that you're actually using. If you are setting infinite timeouts to solve a specific problem, describe that problem.

Yes, this is very helpful!

> 2. What do the tooltips on the fields in the UI say?

**I wanted to highlight this element of your answer -- it is very helpful.** 
Before I post, I always try to find the answer... so helping me understand where to search is great for next time. Often, the questions I'm asking are because I don't know where find the info. In any response, I love it when the person tells me how they found the answer. 

> This may require scrounging through code to validate the behavior for "0". A good search to start with is https://github.com/nextgenhealthcare/connect/search?l=Java&p=2&q=timeout then look for the actual connector implementations in the server or connector code. 

Another great example of teaching me how to fish!

And then there's the idea of searching the Java source code. I could probably sort that out, but I'm not a Java guy... so I was afraid that might be the response. 

> I think this highlights that MC does not have a universal behavior for all timeout fields.

Well, that's problematic.

> Finally - In my opinion an infinite timeout is not a good practice. At least not a good practice without a genuine connection pool manager. Large timeouts are fine, but you eventually want to terminate a connection that is not being used.

Yes, I agree. Really, really agree.
